src/cli/java/org/commcare/util/screen/EntityListSubscreen.java

```diff
@@ -45,24 +49,64 @@ public class EntityListSubscreen extends Subscreen<EntityScreen> {
 
     public EntityListSubscreen(Detail shortDetail, Vector<TreeReference> references, EvaluationContext context,
-            boolean handleCaseIndex) throws CommCareSessionException {
+            boolean handleCaseIndex, EntityScreenContext entityScreenContext) throws CommCareSessionException {
         mHeader = createHeader(shortDetail, context);
         this.shortDetail = shortDetail;
         this.rootContext = context;
         this.entitiesRefs = new TreeReference[references.size()];
         this.handleCaseIndex = handleCaseIndex;
+        this.entityScreenContext = entityScreenContext;
         references.copyInto(entitiesRefs);
         actions = shortDetail.getCustomActions(context);
-        initEntities(context, shortDetail);
+        initEntities(context, shortDetail, entityScreenContext);
     }
 
-    private void initEntities(EvaluationContext context, Detail shortDetail) {
+    private void initEntities(EvaluationContext context, Detail shortDetail,
+            EntityScreenContext entityScreenContext) {
         NodeEntityFactory nodeEntityFactory = new NodeEntityFactory(shortDetail, context);
         entities = new ArrayList<>();
         for (TreeReference reference : entitiesRefs) {
             entities.add(nodeEntityFactory.getEntity(reference));
         }
         nodeEntityFactory.prepareEntities(entities);
+        filterEntities(entityScreenContext);
+        sortEntities(entityScreenContext);
+    }
+
+    private void filterEntities(EntityScreenContext entityScreenContext) {
+        String searchText = entityScreenContext.getSearchText();
+        boolean isFuzzySearchEnabled = entityScreenContext.isFuzzySearch();
+        if (searchText != null && !"".equals(searchText)) {
+            NodeEntityFactory nodeEntityFactory = new NodeEntityFactory(shortDetail, rootContext);
+            EntityStringFilterer filterer = new EntityStringFilterer(searchText.split(" "),
+                    nodeEntityFactory, entities, isFuzzySearchEnabled);
+            entities = filterer.buildMatchList();
+        }
+    }
+
+    private void sortEntities(EntityScreenContext entityScreenContext) {
+        int sortIndex = entityScreenContext.getSortIndex();
+        int[] order;
+        boolean reverse = false;
+        if (sortIndex != 0) {
+            if (sortIndex < 0) {
+                reverse = true;
+                sortIndex = Math.abs(sortIndex);
+            }
+            // sort index is one indexed so adjust for that
+            int sortFieldIndex = sortIndex - 1;
+            order = new int[]{sortFieldIndex};
+        } else {
+            order = shortDetail.getOrderedFieldIndicesForSorting();
+            for (int i = 0; i < shortDetail.getFields().length; ++i) {
+                String header = shortDetail.getFields()[i].getHeader().evaluate();
+                if (order.length == 0 && !"".equals(header)) {
+                    order = new int[]{i};
+                }
+            }
```

Comment:
nit: I think this could be improved by checking order length before the loop and exiting early once a non-blank field is found.

(could leave it for now)

Reply:
Good catch, https://github.com/dimagi/commcare-core/pull/1289/commits/df48f8c50ad6ec67936151745919d1953163e329